instance/models/instance.py

```diff
@@ -17,622 +17,164 @@
+    @name.setter
+    def name(self, new_name):
+        """ Change the 'name' """
+        self.ref.name = new_name
 
     @property
-    def studio_sub_domain(self):
-        """
-        Studio sub-domain name (eg. 'studio.master')
-        """
-        return 'studio.{}'.format(self.sub_domain)
+    def created(self):
+        """ Get this instance's created date, which is stored in the InstanceReference """
+        return self.ref.created
 
     @property
-    def studio_domain(self):
-        """
-        Studio full domain name (eg. 'studio.master.sandbox.opencraft.com')
-        """
-        return '{0.studio_sub_domain}.{0.base_domain}'.format(self)
+    def modified(self):
+        """ Get this instance's modified date, which is stored in the InstanceReference """
+        return self.ref.modified
 
-    @property
-    def studio_url(self):
-        """
-        Studio URL
-        """
-        return u'{0.protocol}://{0.studio_domain}/'.format(self)
+    def save(self, *args, **kwargs):
+        """ Save this Instance """
+        super().save(*args, **kwargs)
+        # Ensure an InstanceReference exists, and update its 'modified' field:
+        if self.ref.instance_id is None:
+            self.ref.instance_id = self.pk  # <- Fix needed when self.ref is accessed before the first self.save()
+        self.ref.save()
 
-    @property
-    def database_name(self):
-        """
-        The database name used for external databases. Escape all non-ascii characters and truncate to 64 chars, the
-        maximum for mysql
+    # pylint: disable=no-member
+    def refresh_from_db(self, using=None, fields=None, **kwargs):
         """
-        name = self.domain.replace('.', '_')
-        allowed = string.ascii_letters + string.digits + '_'
-        escaped = ''.join(char for char in name if char in allowed)
-        return truncate_name(escaped, length=64)
+        Reload from DB, or load related field.
 
-    @property
-    def mysql_database_name(self):
-        """
-        The mysql database name for this instance
+        We override this to ensure InstanceReference is reloaded too.
+        Otherwise, the name/created/modified properties could be out of date, even after
+        Instance.refresh_from_db() is called.
         """
-        return self.database_name
+        if fields is None:
+            self.ref.refresh_from_db()
+        super().refresh_from_db(using=using, fields=fields, **kwargs)
 
-    @property
-    def mysql_database_names(self):
-        """
-        List of mysql database names
+    @staticmethod
+    def on_post_save(sender, instance, created, **kwargs):
         """
-        return [self.mysql_database_name]
+        Called when an instance is saved.
 
-    @property
-    def mongo_database_name(self):
-        """
-        The name of the main external mongo database
+        Each Instance subclass must explicitly connect() to this event in its models.py.
```

Comment:
@smarnach I agree with your line of reasoning, though I think the main reason that signals are used instead of `save()` for something like this is so that the save will succeed even if publishing the update to swampdragon fails.

I like this cleanup though and think the simplicity of just using `save()` is better. I will post a follow-up PR (part of this same ticket) since I really want to get this thing landed.